### ISSUE TITLE
Wireguard tunnels fail to start with empty DNS cache.

### ISSUE THREAD

**Issue by weedy** [celzero/rethink-app#1533]
I can repro this best rebooting my phone. 

As far as I can tell there's some timing weirdness at boot where it doesn't resolve the endpoint domain before setting up the tunnel.

And since I use "Always on", all other DNS queries fail and I have no connectivity to the internet.

Closing and opening rethink fixes everything and it never happens again unless I reboot.

DNS: DoT
Advanced DNS: no
DNS booster: yes
Never proxy DNS: no

Proxy: Warp, always-on

Loopback proxy: no
Do not local IPs: yes
IP version: auto
Connectivity checks: yes

**Comment by ignoramous** [comment]
Sounds terrible.

I am unable to reproduce the issue you're seeing.

Does your WireGuard not setup an IP at all for *Peer Endpoint*? Not being able to resolve IPs for an endpoint is an unrecoverable failure (ie, the WireGuard tunnel isn't even started).

Are you technical enough? If so, cann you capture output from `adb logcat`? Make sure to put Rethink in *Very Verbose* mode from *Configure* -> *Settings* -> *Log level*.

> Closing and opening rethink fixes everything

The next time, will you please see if tapping on the "refresh" button at the top right-hand corner in *Configure* -> *Proxy* UI fixes it?

> And since I use "Always on", all other DNS queries fail and I have no connectivity to the internet.

You can disable, if you so wish, this behaviour by turning ON *Never proxy DNS in *Configure* -> *DNS*.

> Cache

Cache isn't used when resolving WireGuard *Peer endpoints*.

**Comment by weedy** [comment]
![Screenshot_20240611_201734_com.celzero.bravedns.jpg](https://github.com/celzero/rethink-app/assets/648786/e1ed7a82-a21d-4654-be2f-f78b3b31a2b9)

I can get logcat, but it will have to be later when I have my real laptop. 



So I can get more info by turning on airplane mode and playing with Rethink. Then I get a toast from Rethink and no change in brokenness.

Refresh button doesn't help. 

Never proxy DNS gets me lookups, but the tunnel still don't repair themselves. So I'm wrong about system DNS cache/resolver.

I was referring to system DNS cache, but doesn't matter I was wrong on that. 


Assuming this is the same error I guess I don't have fully connected network at the time of Rethink launch and the interfaces aren't coming up properly. 

No amount of flipping airplane mode or switching from LTE to WiFi and back clears it. I have to Stop/Start or turn off always on. 
Without Always on I haven't configured any apps for my main VPN tunnel, it seems up pop up then.

**Comment by ignoramous** [comment]
Looks like, like I mentioned above, not being able to resolve a *Peer Endpoint* domain name is not a recoverable failure (even with "refresh"). I'll try to make it recoverable with "refresh", but no guarantees.

As for why it consistently happens after a reboot is baffling. Logs would help understand just what's going on.

**Comment by soshial** [comment]
When I try to set log buffer to 16M, ADB returns error "MAX log buffer size is 5 MiB. So set it to 5 MiB"

**Comment by weedy** [comment]
Can we cache the last answer from the resolver?

If the lookup fails try the last IP that worked before dying?

**Comment by ignoramous** [comment]
> Can we cache the last answer from the resolver?

Yes, but DNS is cheap and isn't worth caching answers to disk, though we may yet implement this for other reasons. See also: https://github.com/celzero/rethink-app/issues/296#issuecomment-1483646387

**Comment by ignoramous** [comment]
> When I try to set log buffer to 16M, ADB returns error "MAX log buffer size is 5 MiB. So set it to 5 MiB"

@soshial Ah, well :)

---

@weedy believe this issue has been fixed with versions since `v055o`? The current is [`v055y`](https://github.com/celzero/rethink-app/releases/tag/v0.5.5y). Please do test whenever you get the chance. If the bug you originally reported here with "Always-on" *Advanced* mode WireGuard still persist, let me know and I'll reopen this bug. Thanks!
